doc/neps/nep-0057-array-in-array-out.rst

```diff
@@ -0,0 +1,231 @@
+to the corresponding scalar.
+
+.. code:: python
+
+   import numpy as np
+   x = np.asarray(1.)
+   np.isscalar(x + x)  # True (main), False (NEP)
+
+The main impact to users is more predictable results due to improved
+consistency within NumPy, between NumPy and the array API standard, and
+between NumPy and other array libraries. Working with the results of reducing
+functions, in particular, will be easier because return values of any rank
+will support boolean indexing assignment.
+
+There is a secondary impact on performance. On typical hardware, execution
+time of conversion from zero-rank arrays to scalars and elementary arithmetic
+operations involving only scalars is on the order of tens of nanoseconds,
+whereas operations involving only zero-rank arrays is on the order of hundreds
+of nanoseconds. Consequently, some elementary arithmetic calculations will be
+slower. On the other hand, conversion from scalars to zero-rank arrays takes a
+few hundred nanoseconds, and many operations, such such as ufuncs and
+operations involving both scalars and rank-zero arrays require conversion from
+scalars to zero-rank arrays. These operations will be faster. We will not
+speculate as to whether this will have a net positive or net negative impact on
+user applications, but the net impact is expected to be small since impact on
+downstream library test suites has been minimal in testing.
+
+Backward compatibility
+----------------------
+
+The motivation of this proposal is to eliminate the surprises associated with
+a scalar result when a zero-rank array would be expected. However, existing
+code may rely on the current behavior, and this presents backward compatibility
+concerns.
+
+The main concern for user code is that the mutable zero-rank arrays that
+replace immutable scalars are no longer hashable. For instance, they cannot
+by used directly as keys of dictionaries, the argument of an ``lru_cache``
+-decorated function, etc. In all circumstances, tbe patch is simple: convert
+the zero-rank array to a scalar with empty-tuple indexing.
+
+Running the test suites of dependent libraries against a branch of NumPy that
+implements these changes has revealed a few other issues. <Library maintainers,
+please summaries these issues here.>
+
+Detailed description
+--------------------
+
+The new functionality will be used in much the same was as old functionality,
+except that sometimes authors will need to convert zero-rank arrays to scalars
+rather than converting scalars to zero-rank arrays. For instance, consider the
+following:
+
+.. code:: python
+
+   import numpy as np
+   rng = np.random.default_rng(85878653462722874072976519960992129768)
+   x = rng.standard_normal(size=10)
+   y = np.sum(x, axis=-1)
+   z = {y: 'a duck'}  # use scalar result as dictionary key
+   y = np.asarray(y)  # convert to array to allow mutation
+   y[y < 0] = np.nan
+
+The use of ``x`` as a dictionary key would need to become ``x[()]``, but ``z``
```

Comment:
The `x` and `y` here should both (all 3 cases) be `y`?

Reply:
Oops; I was messing with the code and forgot to update the text. Thanks!